freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/freeipa_replica_install.sh

```diff
@@ -13,7 +13,14 @@ echo "The first ipaddress of the host is $IPADDR"
 if [ ! -f /etc/resolv.conf.orig ]; then
   cp /etc/resolv.conf /etc/resolv.conf.orig
 fi
-FORWARDERS=$(grep -Ev '^#|^;' /etc/resolv.conf.orig | grep nameserver | awk '{print "--forwarder " $2}');
+NAMESERVER=$(grep -Ev '^#|^;' /etc/resolv.conf.orig | grep nameserver | awk '{print $2}')
+
+if [[ "${NAMESERVER%.*.*}" == "169.254" ]]; then
+  echo "IPA does not work with link-local IP addresses"
+  FORWARDERS="--auto-forwarders"
```

Comment:
This basically means that GCP will not be HA. It will only be able to resolve external DNS queries via the original primary gateway. If that node is stopped, experiencing an outage, or has been repaired, then it will fail. In the case of rebuilding the primary gateway that failure will be permanent.

I don't know what we do in this case, but this will result in the customer being unable to access external data sources after a repair.

Is it possible to configure the GCP DNS servers to be allocated an IP address that is not link-local?

Reply:
I'm worried about bringing patches to freeipa and maintaining our own distribution in the end :)

Reply:
All great points. How about I set the forwarder to `8.8.8.8` in the replica.

Few things,

1. The experiences for GCP are planned for H2 2021.
2. The datahub and datalake nodes do not need FreeIPA to resolve external DNS addresses as they have unbound.
3. Patching is a good option, but as @lacikaaa mentioned there might be maintenance concerns. Also it will take time to do it cleanly for an extended period of maintenance.

@lacikaaa @jamisonbennett please let me know what you think.

Reply:
@cegganesh84 so the primary instance creation succeeds and it has the right DNS server configured. Is this statement right?
If it's right, could we configure the forwarder for the replicas after the installation?

Reply:
Good question. No, the primary has auto-forwarder only. Which it will point to the global DNS via the /etc/resolv.conf. This is the problem only for the replicas. Replica's /etc/resolv.conf should point to the primary except for failover scenarios. @jamisonbennett please correct me here if I am making too many bad assumptions.

Reply:
> This is the problem only for the replicas.
Yes

> Replica's /etc/resolv.conf should point to the primary except for failover scenarios.
No, the replicas should always reference the upstream DNS server. The primary instance and replicas should all identical DNS records since that data is stored in LDAP and replicated across all the nodes. When DNS resolution is external to FreeIPA (e.g. www.google.com), then we want to resolve using the upstream DNS servers.

> How about I set the forwarder to 8.8.8.8 in the replica.
While this would resolve public entries, I fear it may miss DNS records for other nodes in the VPC that are outside of FreeIPA's zone. It may also miss other entries.

Reply:
> The primary instance and replicas should all identical DNS records since that data is stored in LDAP and replicated across all the nodes.

How about setting the forwarder to 127.0.0.1. That should point to /etc/resolv.conf anyways.

Reply:
> How about setting the forwarder to 127.0.0.1. That should point to /etc/resolv.conf anyways.

That would mean the ability to run unbound in a FreeIPA node as well, which is not possible today.

> The primary instance and replicas should all identical DNS records since that data is stored in LDAP and replicated across all the nodes.

@jamisonbennett I will go back to this. What purpose does /etc/resolv.conf serve in FreeIPA replica? In other words how --auto-forwarders work differently in primary vs replica?

Reply:
https://www.systutorials.com/docs/linux/man/1-ipa-replica-install/
```
--forwarder=IP_ADDRESS
Add a DNS forwarder to the DNS configuration. You can use this option multiple times to specify more forwarders, but at least one must be provided, unless the --no-forwarders option is specified.
--no-forwarders
Do not add any DNS forwarders. Root DNS servers will be used instead.
--auto-forwarders
Add DNS forwarders configured in /etc/resolv.conf to the list of forwarders used by IPA DNS.
```

@jamisonbennett It looks like we want the behavior of `--auto-forwarders` only in the replica too. So there is no need for explicitly setting to the one in /etc/resolv.conf.orig, we could simply swap the files in the replica for GCP.

Reply:
I think the confusion was because of how the DNS is bootstrapped CB-8271.

Because of which the primary's /etc/resolv.conf is pointing to the cloud provider DNS and the replica's /etc/resolv.conf is pointing to the primary's ip-address.

Reply:
you can't as the replica must find the primary during installation, and that works only if the primary is set as DNS server

Reply:
I think we have the following options:
1. Revert CB-8271 and use `--auto-forwarders`
CB-8271 allowed every instance of FreeIPA to no longer have a /etc/hosts which pointed to every other instance of FreeIPA. It was a much cleaner bootstrapping, especially when you consider a cluster that had many repairs. There are some potential problems with having lots of old entries in /etc/hosts, if we are somehow still referencing the hostname somewhere it will attempt to go to an IP address that may be from a different cluster of FreeIPA (e.g. if the IP was reused) rather than getting a clear error that the host does not exist. We may need to be more vigilant in cleaning up /etc/hosts.

> you can't as the replica must find the primary during installation, and that works only if the primary is set as DNS server

I think it may actually be possible if:
* all the FreeIPA servers setup /etc/hosts with every FreeIPA servers IP address during provisioning and repair
* replicas don't join the cluster first but the are just installed as a part of the cluster
* we add some cleanup for all FreeIPA instances /etc/hosts after repair

But this still has a downside. CB-8356 becomes less possible. I don't know how we will be able to update the DNS forwarders after installation to 169.x.x.x IP addresses unless FreeIPA is patched to support it.

2. Patch FreeIPA via salt to work with the GCP IP addresses for DNS forwarding.
It looked like FreeIPA possibly fixed this for newer versions.

Reply:
I think the option to revert CB-8271 and use `--auto-forwarders` is the quickest solution but patching FreeIPA to work with the GCP IP addresses is a better technical (assuming the patch actually works cleanly against our version of FreeIPA, I didn't test it). I know there has been concerns about patching FreeIPA, so I am good with either option.

Reply:
Thanks for the input folks. I took a slightly different approach. Wanted to communicate in terms of code this time. Please review the latest patch.

Reply:
I actually tested this latest change and it worked.

The replica install succeeded.

The forwarders are used correctly `2021-01-29T06:20:05Z DEBUG will use DNS forwarders: [CheckedIPAddressLoopback('10.114.16.44'), '169.254.169.254']`

It automagically changes the resolv.conf to point to itself after that,

```
2021-01-29T06:28:44Z DEBUG   [8/8]: changing resolv.conf to point to ourselves
2021-01-29T06:28:44Z DEBUG Backing up system configuration file '/etc/resolv.conf'

```

<img width="1513" alt="Screen Shot 2021-01-28 at 10 36 12 PM" src="https://user-images.githubusercontent.com/18580046/106240624-71249900-61b9-11eb-8f47-534d76260d4c.png">